### DISCUSSION TITLE
Tasks failing often with no logs

### DISCUSSION THREAD

**Discussion by zorzigio** [apache/airflow#21624] [General]
### Apache Airflow version

2.2.3 (latest released)

### What happened

I often have tasks failing in Airflow and no logs are produced.
If I clear the tasks, it will then run successfully. 

Some tasks are stuck in queue state and even if cleared, will get stuck again in queue state.

Happy to provide more details if needed.

### What you expected to happen

Tasks running successfully

### Operating System

Debian GNU/Linux 10 (buster)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==2.4.0
apache-airflow-providers-celery==2.1.0
apache-airflow-providers-cncf-kubernetes==2.2.0
apache-airflow-providers-docker==2.3.0
apache-airflow-providers-elasticsearch==2.1.0
apache-airflow-providers-ftp==2.0.1
apache-airflow-providers-google==6.2.0
apache-airflow-providers-grpc==2.0.1
apache-airflow-providers-hashicorp==2.1.1
apache-airflow-providers-http==2.0.1
apache-airflow-providers-imap==2.0.1
apache-airflow-providers-microsoft-azure==3.4.0
apache-airflow-providers-mysql==2.1.1
apache-airflow-providers-odbc==2.0.1
apache-airflow-providers-postgres==2.4.0
apache-airflow-providers-redis==2.0.1
apache-airflow-providers-sendgrid==2.0.1
apache-airflow-providers-sftp==2.3.0
apache-airflow-providers-slack==4.1.0
apache-airflow-providers-sqlite==2.0.1
apache-airflow-providers-ssh==2.3.0

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

I am using the Celery Executor with KEDA enables on Kubernetes. The nodepool is set on autoscaling.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by zorzigio** [comment]
Not sure if this is related, but I am also seeing an "Import error" every now and then, apparently appearing out of thin air

```bash
Broken DAG: [/opt/airflow/dags/repo/dags/documents/DEV/file_pipeline.py] Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 574, in serialize_operator
    serialize_op['params'] = cls._serialize_params_dict(op.params)
  File "/home/airflow/.local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 447, in _serialize_params_dict
    if f'{v.__module__}.{v.__class__.__name__}' == 'airflow.models.param.Param':
AttributeError: 'str' object has no attribute '__module__'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 935, in to_dict
    json_dict = {"__version": cls.SERIALIZER_VERSION, "dag": cls.serialize_dag(var)}
  File "/home/airflow/.local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 847, in serialize_dag
    raise SerializationError(f'Failed to serialize DAG {dag.dag_id!r}: {e}')
airflow.exceptions.SerializationError: Failed to serialize DAG 'DEV_file_pipeline': 'str' object has no attribute '__module__'
```

This disappears when I refresh the page

**Comment by chenglongyan** [comment]
related: #20648

**Reply by potiuk** [reply]
There are some edge-cases that will likely be handled in 2.2.4 which is ALMOST out. I do not know all the details - maybe @ephraimbuddy will be able to confirm if this looks like one of those propblems already addressed there or (I believe) there is one we are still unsure if the solution we have is working, @ephraimbuddy - maybe this is a good idea to involve @zorzigio and if we see some similarities with solved problems @zorzigio could help with applying some patches and seing if they help in his case ?

**Comment by potiuk** [comment]
Some logs should be present - can you also check if you have enough resources / kubernetes logs ? It is lilkely your tasks are killed due to lack of resources (memory most likely).

**Reply by zorzigio** [reply]
It seems to me that the pods have enough resources

```
NAME↑                                 READY   RESTARTS STATUS     CPU    MEM   %CPU/R    %CPU/L    %MEM/R   %MEM/L
airflow-flower-5c9b79c65b-dgcvx       1/1            0 Running     13    181      n/a       n/a       n/a      n/a
airflow-redis-0                       1/1            0 Running      1      8      n/a       n/a       n/a      n/a
airflow-scheduler-6f97f58856-gc592    3/3            0 Running    453    993       45       n/a        48      n/a
airflow-statsd-75ff45fcbc-dp7tn       1/1            0 Running      2     62      n/a       n/a       n/a      n/a
airflow-triggerer-867d8cb446-2kcg4    2/2            0 Running     68    989      n/a       n/a       n/a      n/a
airflow-webserver-f456bdfc-pq2kt      1/1            0 Running      3   1066        0         0        52       52
airflow-worker-df54f849-jjqfr         2/2            1 Running     10   1727        1       n/a        84      n/a
```

Also looking at the Kubernetes nodes everything looks fine to me in terms of resources

```
NAME                       CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%
aks-nodepool1-vmss000000   216m         11%    3978Mi          87%
aks-nodepool1-vmss000006   86m          4%     3413Mi          74%
aks-nodepool1-vmss00000f   156m         8%     2939Mi          64%
aks-nodepool2-vmss000000   243m         12%    7258Mi          57%
aks-nodepool2-vmss00000e   55m          2%     2332Mi          18%
aks-nodepool2-vmss00000f   274m         14%    1396Mi          11%
aks-nodepool2-vmss00000g   60m          3%     2867Mi          22%
```

**Comment by potiuk** [comment]
Converting to discussion until more information provided.

**Comment by zorzigio** [comment]
I am also seeing an increase in the scheduler memory usage over time

![scheduler](https://user-images.githubusercontent.com/17679872/154487896-8dae7aec-31a2-4cd9-91fa-a82d96811ece.png)

**Comment by potiuk** [comment]
Which version of Airlfow do you use @zorzigio ? 

From your observation about  memory I assume it is before 2.1.4? 

Is it "active" or "cache" memory ? If it's cache, this is harmless and expected (In 2.1.4 https://github.com/apache/airflow/pull/18054 we added kerne advisory to not cache log file memory). The cache growth was from log writing, but cache memory growth is entirely harmless and expected in earlier versions - the memory will be freed as needed and it's normal behaviour of Unix.

Before we dive deeper I strongly advise (if my guess is right) to upgrade to latest release of Airlfow. There were already a number of fixes since then so this is very likely.

See this thread: https://github.com/apache/airflow/discussions/21499#discussioncomment-2186204 if you want to see testimony of others who did.

**Comment by zorzigio** [comment]
@potiuk, thanks for looking into this.

I am using the latest version of Airflow (which of today is 2.2.3)

Here is the output from  `airflow info`

```
Apache Airflow
version                | 2.2.3
executor               | CeleryExecutor
task_logging_handler   | airflow.utils.log.file_task_handler.FileTaskHandler
sql_alchemy_conn       | postgresql://REDACTED:REDACTED@postgresql.postgresql.svc.cluster.local:5432/airflow_db?sslmode=disable
dags_folder            | /opt/airflow/dags/repo/dags
plugins_folder         | /opt/airflow/plugins
base_log_folder        | /opt/airflow/logs
remote_base_log_folder |


System info
OS              | Linux
architecture    | x86_64
uname           | uname_result(system='Linux', node='airflow-scheduler-6f97f58856-gc592', release='5.4.0-1047-azure', version='#49~18.04.1-Ubuntu SMP Thu Apr 22 21:28:54 UTC 2021', machine='x86_64')
locale          | ('en_US', 'UTF-8')
python_version  | 3.9.9 (main, Dec 21 2021, 10:35:05)  [GCC 8.3.0]
python_location | /usr/local/bin/python


Tools info
git             | NOT AVAILABLE
ssh             | OpenSSH_7.9p1 Debian-10+deb10u2, OpenSSL 1.1.1d  10 Sep 2019
kubectl         | NOT AVAILABLE
gcloud          | NOT AVAILABLE
cloud_sql_proxy | NOT AVAILABLE
mysql           | mysql  Ver 8.0.28 for Linux on x86_64 (MySQL Community Server - GPL)
sqlite3         | 3.27.2 2019-02-25 16:06:06 bd49a8271d650fa89e446b42e513b595a717b9212c91dd384aab871fc1d0alt1
psql            | psql (PostgreSQL) 11.14 (Debian 11.14-0+deb10u1)


Paths info
airflow_home    | /opt/airflow
system_path     | /home/airflow/.local/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
python_path     | /home/airflow/.local/bin:/opt/airflow/dags/repo:/usr/local/lib/python39.zip:/usr/local/lib/python3.9:/usr/local/lib/python3.9/lib-dynload:/home/airflow/.local/lib/python3.9/site-packages:/usr/local/lib/python3.9/site-packages:/opt/airflow/dags/repo/dags:/opt/airflow/config:/opt/airflow/plugins
airflow_on_path | True


Providers info
apache-airflow-providers-amazon          | 2.4.0
apache-airflow-providers-celery          | 2.1.0
apache-airflow-providers-cncf-kubernetes | 2.2.0
apache-airflow-providers-docker          | 2.3.0
apache-airflow-providers-elasticsearch   | 2.1.0
apache-airflow-providers-ftp             | 2.0.1
apache-airflow-providers-google          | 6.2.0
apache-airflow-providers-grpc            | 2.0.1
apache-airflow-providers-hashicorp       | 2.1.1
apache-airflow-providers-http            | 2.0.1
apache-airflow-providers-imap            | 2.0.1
apache-airflow-providers-microsoft-azure | 3.4.0
apache-airflow-providers-mysql           | 2.1.1
apache-airflow-providers-odbc            | 2.0.1
apache-airflow-providers-postgres        | 2.4.0
apache-airflow-providers-redis           | 2.0.1
apache-airflow-providers-sendgrid        | 2.0.1
apache-airflow-providers-sftp            | 2.3.0
apache-airflow-providers-slack           | 4.1.0
apache-airflow-providers-sqlite          | 2.0.1
apache-airflow-providers-ssh             | 2.3.0


```

**Reply by potiuk** [reply]
Can you check the memory type that is growing anyway ? It might some of your tasks that are allocatng cache (basically any time you write to a file in UNIX if you don't advise kernel, it will pages of maped memory it to the cache so - memory growth (as long as it is cache) is natural thing for UNIX. I'd like to exclude that (See https://github.com/apache/airflow/pull/18054 for details on how to check it)

**Reply by zorzigio** [reply]
It is not a cache memory related issue. 

I was reading your comment [here](https://github.com/apache/airflow/issues/14924#issuecomment-911812067) and you suggest to use `container_memory_working_set_bytes` rather than `container_memory_cache`.

Here is another graph from Grafana where the prefix stands for:
- cmwsb_ = `container_memory_working_set_bytes`
- cmc_ = `container_memory_cache`

![scheduler](https://user-images.githubusercontent.com/17679872/154570036-a6f2322a-9c8a-411c-84f0-ffb97acb6ba6.png)

I tried to drop the cache as a test anyway, but I get:

```bash
> echo 1 > /proc/sys/vm/drop_caches
bash: /proc/sys/vm/drop_caches: Read-only file system
```

**Reply by potiuk** [reply]
That would suggest some problem with your DAGS - the memory growing issue has been fixed and we have not seen anyone else reporting it - do your dags use some libraries that  can allocate memory (some C/C++ libraries/custom code/libraries - this is likely a memory leak in some of this code? Those libraries could also create unlogged failures of Airflow tasks if they sigsegv randomly for example - do you have some custom C/C++ code or heavy use some of C/C++ or other low-language libraries ?

**Reply by zorzigio** [reply]
Not much going on in the Dockerfile used for Airflow

```dockerfile
FROM apache/airflow:2.2.3-python3.9


# https://airflow.apache.org/docs/docker-stack/build.html
USER root
RUN apt-get update \
  && apt-get install -y --no-install-recommends \
         libmagic-dev \
  && apt-get autoremove -yqq --purge \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
USER airflow

RUN pip install --no-cache-dir --user \
    azure-storage-queue==12.1.6 \
    apache-airflow-providers-cncf-kubernetes \
    azure-keyvault-secrets azure-identity \
    XlsxWriter pandas \
    opencv-python-headless imutils \
    azure-cognitiveservices-vision-computervision python-magic \
    pydantic==1.9.0
```

**Reply by Kirgod** [reply]
@zorzigio any luck with that issue? Literally same picture, `container_memory_working_set_bytes` is growing and growing, from 200MB to 4Gi for 3 days.

**Comment by ephraimbuddy** [comment]
I think the issue you are having is about DAG serialization. You can try clearing the serialized DAGs. There's an `airflow db reserialize` command but it's not released yet. Try this:
```python
from airflow.models.serialized_dag import SerializedDagModel
from airflow.settings import Session
session = Session()
session.query(SerializedDagModel).delete()
session.commit()
```
To have airflow reserialize the dags again

When there is an issue running a task https://github.com/apache/airflow/blob/5a38d15e539ca3bb16a86d14a6da2939f1ebb257/airflow/cli/commands/task_command.py#L331 due to serialization, airflow may not have logs at that time. Here's where it's handled by the scheduler: https://github.com/apache/airflow/blob/5a38d15e539ca3bb16a86d14a6da2939f1ebb257/airflow/jobs/scheduler_job.py#L613-L63

**Reply by zorzigio** [reply]
Thanks @ephraimbuddy for the suggestion.
I tried re-serialising the DAGs, but the problem persists

**Reply by zorzigio** [reply]
Does the DAG serialisation create any logs at all that I could look at to see what is going on?
Also, why would the webserver show the message about the error serialisation problem and then disappear once the page is refreshed?
Shouldn't the serialisation problem be a problem all the times rather than disappearing on its own?

**Reply by potiuk** [reply]
I think you should migrate to - just released 2.2.4 and see if the problems are fixed.

**Reply by zorzigio** [reply]
Thanks @potiuk. 
I migrated to 2.2.4. After this migration I can see that the tasks that were stuck in "running", are not stuck anymore.
I will keep an eye on the other issues

**Reply by potiuk** [reply]
This is cool. Do let us know. I think 2.2.4 might really nail a few final problems with queeuing  (@ephraimbuddy  - I think that's mostly your great job :) )

**Comment by zorzigio** [comment]
This issue seems similar to #21082

**Comment by dharmendrakariya** [comment]
Any update on this?

**Reply by ephraimbuddy** [reply]
Does this still happen in 2.5.0

**Reply by dharmendrakariya** [reply]
yes!

**Reply by potiuk** [reply]
If you have reproducible error and some more logs and the ways how it can be triggered - then please submit more information (ideally in the form of reproducible issue) so that we could diagnose it. The fact of life is that if we have not enough information, circumstances. logs, guesses, and some results from your investigation which would be reproducible or at least hint at some directions, there is very little anyone can do about it and we do not have a lot of time as maintainers to try to guess and run multiple tests there - this is an open-source project where we contribute and solve similar problems mostly in our free tiem. So we have to rely on our users doing a lot of investigation and providing some ways of reproducibility. So getting more information and investigations from users experiencing it is the only way any diagnosis and fixes can be sped-up.

**Reply by Alaeddine22** [reply]
I might have reproduced the same behavior. Please check https://github.com/apache/airflow/discussions/33665 for more details.
